File: app/src/main/java/com/nextcloud/talk/adapters/ParticipantDisplayItem.java
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2022 Daniel Calviño Sánchez <danxuliu@gmail.com>
 * SPDX-FileCopyrightText: 2021 Marcel Hibbe <dev@mhibbe.de>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.adapters;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;

import com.nextcloud.talk.call.CallParticipantModel;
import com.nextcloud.talk.call.RaisedHand;
import com.nextcloud.talk.models.json.participants.Participant;
import com.nextcloud.talk.utils.ApiUtils;
import com.nextcloud.talk.utils.DisplayUtils;

import org.webrtc.EglBase;
import org.webrtc.MediaStream;
import org.webrtc.PeerConnection;

public class ParticipantDisplayItem {

    /**
     * Shared handler to receive change notifications from the model on the main thread.
     */
    private static final Handler handler = new Handler(Looper.getMainLooper());

    private final ParticipantDisplayItemNotifier participantDisplayItemNotifier = new ParticipantDisplayItemNotifier();

    private final Context context;

    private final String baseUrl;
    private final String defaultGuestNick;
    private final EglBase rootEglBase;

    private final String session;
    private final String streamType;

    private final String roomToken;

    private final CallParticipantModel callParticipantModel;

    private Participant.ActorType actorType;
    private String actorId;
    private String userId;
    private PeerConnection.IceConnectionState iceConnectionState;
    private String nick;
    private String urlForAvatar;
    private MediaStream mediaStream;
    private boolean streamEnabled;
    private boolean isAudioEnabled;
    private RaisedHand raisedHand;

    public interface Observer {
        void onChange();
    }

    private final CallParticipantModel.Observer callParticipantModelObserver = new CallParticipantModel.Observer() {
        @Override
        public void onChange() {
            updateFromModel();
        }

        @Override
        public void onReaction(String reaction) {
        }
    };

    public ParticipantDisplayItem(Context context, String baseUrl, String defaultGuestNick, EglBase rootEglBase,
                                  String streamType, String roomToken, CallParticipantModel callParticipantModel) {
        this.context = context;

        this.baseUrl = baseUrl;
        this.defaultGuestNick = defaultGuestNick;
        this.rootEglBase = rootEglBase;

        this.session = callParticipantModel.getSessionId();
        this.streamType = streamType;

        this.roomToken = roomToken;

        this.callParticipantModel = callParticipantModel;
        this.callParticipantModel.addObserver(callParticipantModelObserver, handler);

        updateFromModel();
    }

    public void destroy() {
        this.callParticipantModel.removeObserver(callParticipantModelObserver);
    }

    private void updateFromModel() {
        actorType = callParticipantModel.getActorType();
        actorId = callParticipantModel.getActorId();
        userId = callParticipantModel.getUserId();
        nick = callParticipantModel.getNick();

        this.updateUrlForAvatar();

        if ("screen".equals(streamType)) {
            iceConnectionState = callParticipantModel.getScreenIceConnectionState();
            mediaStream = callParticipantModel.getScreenMediaStream();
            isAudioEnabled = true;
            streamEnabled = true;
        } else {
            iceConnectionState = callParticipantModel.getIceConnectionState();
            mediaStream = callParticipantModel.getMediaStream();
            isAudioEnabled = callParticipantModel.isAudioAvailable() != null ?
                callParticipantModel.isAudioAvailable() : false;
            streamEnabled = callParticipantModel.isVideoAvailable() != null ?
                callParticipantModel.isVideoAvailable() : false;
        }

        raisedHand = callParticipantModel.getRaisedHand();

        participantDisplayItemNotifier.notifyChange();
    }

    private void updateUrlForAvatar() {
        if (actorType == Participant.ActorType.FEDERATED) {
            int darkTheme = DisplayUtils.INSTANCE.isDarkModeOn(context) ? 1 : 0;
            urlForAvatar = ApiUtils.getUrlForFederatedAvatar(baseUrl, roomToken, actorId, darkTheme, true);
        } else if (!TextUtils.isEmpty(userId)) {
            urlForAvatar = ApiUtils.getUrlForAvatar(baseUrl, userId, true);
        } else {
            urlForAvatar = ApiUtils.getUrlForGuestAvatar(baseUrl, getNick(), true);
        }
    }

    public boolean isConnected() {
        return iceConnectionState == PeerConnection.IceConnectionState.CONNECTED ||
            iceConnectionState == PeerConnection.IceConnectionState.COMPLETED ||
            // If there is no connection state that means that no connection is needed, so it is a special case that is
            // also seen as "connected".
            iceConnectionState == null;
    }

    public String getNick() {
        if (TextUtils.isEmpty(userId) && TextUtils.isEmpty(nick)) {
            return defaultGuestNick;
        }

        return nick;
    }

    public String getUrlForAvatar() {
        return urlForAvatar;
    }

    public MediaStream getMediaStream() {
        return mediaStream;
    }

    public boolean isStreamEnabled() {
        return streamEnabled;
    }

    public EglBase getRootEglBase() {
        return rootEglBase;
    }

    public boolean isAudioEnabled() {
        return isAudioEnabled;
    }

    public RaisedHand getRaisedHand() {
        return raisedHand;
    }

    public Participant.ActorType getActorType() {
        return actorType;
    }

    public void addObserver(Observer observer) {
        participantDisplayItemNotifier.addObserver(observer);
    }

    public void removeObserver(Observer observer) {
        participantDisplayItemNotifier.removeObserver(observer);
    }

    @Override
    public String toString() {
        return "ParticipantSession{" +
                "userId='" + userId + '\'' +
                ", actorType='" + actorType + '\'' +
                ", actorId='" + actorId + '\'' +
                ", session='" + session + '\'' +
                ", nick='" + nick + '\'' +
                ", urlForAvatar='" + urlForAvatar + '\'' +
                ", mediaStream=" + mediaStream +
                ", streamType='" + streamType + '\'' +
                ", streamEnabled=" + streamEnabled +
                ", rootEglBase=" + rootEglBase +
                ", raisedHand=" + raisedHand +
                '}';
    }
}

File: app/src/main/java/com/nextcloud/talk/adapters/ParticipantDisplayItem.kt
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2022 Daniel Calviño Sánchez <danxuliu@gmail.com>
 * SPDX-FileCopyrightText: 2021-2025 Marcel Hibbe <dev@mhibbe.de>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.adapters

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.Log
import android.view.ViewGroup
import com.nextcloud.talk.call.CallParticipantModel
import com.nextcloud.talk.call.RaisedHand
import com.nextcloud.talk.models.json.participants.Participant.ActorType
import com.nextcloud.talk.utils.ApiUtils.getUrlForAvatar
import com.nextcloud.talk.utils.ApiUtils.getUrlForFederatedAvatar
import com.nextcloud.talk.utils.ApiUtils.getUrlForGuestAvatar
import com.nextcloud.talk.utils.DisplayUtils.isDarkModeOn
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.webrtc.EglBase
import org.webrtc.MediaStream
import org.webrtc.PeerConnection.IceConnectionState
import org.webrtc.SurfaceViewRenderer

data class ParticipantUiState(
    val sessionKey: String,
    val nick: String,
    val isConnected: Boolean,
    val isAudioEnabled: Boolean,
    val isStreamEnabled: Boolean,
    val raisedHand: Boolean,
    val avatarUrl: String?,
    val mediaStream: MediaStream?
)

@Suppress("LongParameterList")
class ParticipantDisplayItem(
    private val context: Context,
    private val baseUrl: String,
    private val defaultGuestNick: String,
    val rootEglBase: EglBase,
    private val streamType: String,
    private val roomToken: String,
    private val callParticipantModel: CallParticipantModel
) {
    private val participantDisplayItemNotifier = ParticipantDisplayItemNotifier()

    private val _uiStateFlow = MutableStateFlow(buildUiState())
    val uiStateFlow: StateFlow<ParticipantUiState> = _uiStateFlow.asStateFlow()

    private val session: String = callParticipantModel.sessionId

    var actorType: ActorType? = null
        private set
    private var actorId: String? = null
    private var userId: String? = null
    private var iceConnectionState: IceConnectionState? = null
    var nick: String? = null
        get() = (if (TextUtils.isEmpty(userId) && TextUtils.isEmpty(field)) defaultGuestNick else field)

    var urlForAvatar: String? = null
        private set
    var mediaStream: MediaStream? = null
        private set
    var isStreamEnabled: Boolean = false
        private set
    var isAudioEnabled: Boolean = false
        private set
    var raisedHand: RaisedHand? = null
        private set
    var surfaceViewRenderer: SurfaceViewRenderer? = null

    val sessionKey: String
        get() = "$session-$streamType"

    interface Observer {
        fun onChange()
    }

    private val callParticipantModelObserver: CallParticipantModel.Observer = object : CallParticipantModel.Observer {
        override fun onChange() {
            updateFromModel()
        }

        override fun onReaction(reaction: String) {
            // unused
        }
    }

    init {
        callParticipantModel.addObserver(callParticipantModelObserver, handler)

        updateFromModel()
    }

    @Suppress("Detekt.TooGenericExceptionCaught")
    fun destroy() {
        callParticipantModel.removeObserver(callParticipantModelObserver)

        surfaceViewRenderer?.let { renderer ->
            try {
                mediaStream?.videoTracks?.firstOrNull()?.removeSink(renderer)
                renderer.clearImage()
                renderer.release()
                (renderer.parent as? ViewGroup)?.removeView(renderer)
            } catch (e: Exception) {
                Log.w("ParticipantDisplayItem", "Error releasing renderer", e)
            }
        }
        surfaceViewRenderer = null
    }

    private fun updateFromModel() {
        actorType = callParticipantModel.actorType
        actorId = callParticipantModel.actorId
        userId = callParticipantModel.userId
        nick = callParticipantModel.nick

        updateUrlForAvatar()

        if (streamType == "screen") {
            iceConnectionState = callParticipantModel.screenIceConnectionState
            mediaStream = callParticipantModel.screenMediaStream
            isAudioEnabled = true
            isStreamEnabled = true
        } else {
            iceConnectionState = callParticipantModel.iceConnectionState
            mediaStream = callParticipantModel.mediaStream
            isAudioEnabled = callParticipantModel.isAudioAvailable ?: false
            isStreamEnabled = callParticipantModel.isVideoAvailable ?: false
        }

        raisedHand = callParticipantModel.raisedHand

        if (surfaceViewRenderer == null && mediaStream != null) {
            val renderer = SurfaceViewRenderer(context).apply {
                init(rootEglBase.eglBaseContext, null)
                setEnableHardwareScaler(true)
                setMirror(false)
            }
            surfaceViewRenderer = renderer
            mediaStream?.videoTracks?.firstOrNull()?.addSink(renderer)
        }

        _uiStateFlow.value = buildUiState()
        participantDisplayItemNotifier.notifyChange()
    }

    private fun buildUiState(): ParticipantUiState =
        ParticipantUiState(
            sessionKey = sessionKey,
            nick = nick ?: "Guest",
            isConnected = isConnected,
            isAudioEnabled = isAudioEnabled,
            isStreamEnabled = isStreamEnabled,
            raisedHand = raisedHand?.state == true,
            avatarUrl = urlForAvatar,
            mediaStream = mediaStream
        )

    private fun updateUrlForAvatar() {
        if (actorType == ActorType.FEDERATED) {
            val darkTheme = if (isDarkModeOn(context)) 1 else 0
            urlForAvatar = getUrlForFederatedAvatar(baseUrl, roomToken, actorId!!, darkTheme, true)
        } else if (!TextUtils.isEmpty(userId)) {
            urlForAvatar = getUrlForAvatar(baseUrl, userId, true)
        } else {
            urlForAvatar = getUrlForGuestAvatar(baseUrl, nick, true)
        }
    }

    val isConnected: Boolean
        get() = iceConnectionState == IceConnectionState.CONNECTED ||
            iceConnectionState == IceConnectionState.COMPLETED ||
            // If there is no connection state that means that no connection is needed,
            // so it is a special case that is also seen as "connected".
            iceConnectionState == null

    fun addObserver(observer: Observer?) {
        participantDisplayItemNotifier.addObserver(observer)
    }

    fun removeObserver(observer: Observer?) {
        participantDisplayItemNotifier.removeObserver(observer)
    }

    override fun toString(): String =
        "ParticipantSession{" +
            "userId='" + userId + '\'' +
            ", actorType='" + actorType + '\'' +
            ", actorId='" + actorId + '\'' +
            ", session='" + session + '\'' +
            ", nick='" + nick + '\'' +
            ", urlForAvatar='" + urlForAvatar + '\'' +
            ", mediaStream=" + mediaStream +
            ", streamType='" + streamType + '\'' +
            ", streamEnabled=" + isStreamEnabled +
            ", rootEglBase=" + rootEglBase +
            ", raisedHand=" + raisedHand +
            '}'

    companion object {
        /**
         * Shared handler to receive change notifications from the model on the main thread.
         */
        private val handler = Handler(Looper.getMainLooper())
    }
}
